crates/storage/src/snapshot.rs

```diff
@@ -69,73 +62,81 @@ impl Snapshot {
         &self,
         key: Vec<u8>,
     ) -> Result<(Option<Vec<u8>>, ics23::CommitmentProof)> {
-        let span = Span::current();
-        let snapshot = self.clone();
+        let span = tracing::Span::current();
+        let rocksdb_snapshot = self.0.snapshot.clone();
+        let db = self.0.db.clone();
 
+        let (_, config) = self.0.multistore_cache.config.route_key_bytes(&key);
+        let version = self.substore_version(&config).unwrap_or(u64::MAX);
+
+        let substore = store::substore::SubstoreSnapshot {
+            config,
+            rocksdb_snapshot,
+            version,
+            db,
+        };
         tokio::task::Builder::new()
-            .name("State::get_with_proof")
-            .spawn_blocking(move || {
-                span.in_scope(|| {
-                    let tree = jmt::Sha256Jmt::new(&*snapshot.0);
-                    tree.get_with_ics23_proof(key, snapshot.version())
-                })
-            })?
+            .name("Snapshot::get_with_proof")
+            .spawn_blocking(move || span.in_scope(|| substore.get_with_proof(key)))?
             .await?
     }
 
+    pub fn substore_version(
+        &self,
+        prefix: &Arc<store::substore::SubstoreConfig>,
+    ) -> Option<jmt::Version> {
+        self.0.multistore_cache.get_version(prefix)
+    }
+
     /// Returns the root hash of this `State`.
     ///
     /// If the `State` is empty, the all-zeros hash will be returned as a placeholder value.
     ///
     /// This method may only be used on a clean [`State`] fork, and will error
     /// if [`is_dirty`] returns `true`.
-    pub async fn root_hash(&self) -> Result<crate::RootHash> {
-        let span = Span::current();
-        let snapshot = self.clone();
+    pub async fn prefix_root_hash(&self, prefix: &str) -> Result<crate::RootHash> {
+        let span = tracing::Span::current();
+        let rocksdb_snapshot = self.0.snapshot.clone();
+        let db = self.0.db.clone();
+
+        let config = self
+            .0
+            .multistore_cache
+            .config
+            .find_substore(prefix.as_bytes());
+
+        // If a substore is not found, we default to the main store.
```

Comment:
I was initially confused by this comment, I think it would be clearer if it read

> If a substore is not found, find_substore will default to the main store

to clarify what exactly is doing the defaulting (it's the part of the code that returns `config`)